Add 'omp end do' statement

EmilyBourne opened the pull request:
- Rename `nowait_expr` in `For` to the more general `end_annotation`
- Remove `_print_OMP_X` functions for which the class `OMP_X` does not exist
- Add `!$omp end do` statement to the end of all OpenMP for loops in Fortran
- Ignore any `#$omp end for` statements provided by the user and raise a warning indicating that the clause is ignored
- Ensure that the fst of AnnotatedComments is set

nhamidn commented:
what if the user specified the `#$omp end for `in the python code for example : 
```
     #$omp parallel num_threads(n)
     #$omp for
     for i in range(0, 10):
         print(i)
     #$omp end for
     #$omp end parallel
```
this will generate fortran  code with multiple end statement : 
```
    !$omp parallel num_threads(n)
    !$omp do
    do i = 0_i64, 9_i64, 1_i64
      print *, i
    end do
    !$omp end do
    !$omp end do
    !$omp end parallel
```
and generate C code with the end statement but in C we don't open the bracket the #$omp for construct:
```
     #pragma omp parallel num_threads(n)
     {
         #pragma omp for
         for (i = 0; i < 10; i += 1)
         {
             printf("%ld\n", i);
         }
     }
 }
```

i think we must ignore any `#$omp end for` pragma since we add our own.

yguclu commented:
@EmilyBourne In the PR description you say "Ignore any !$omp end do statements provided by the user", but I am not sure that I understand what you mean.

Since we changed our syntax from `!$omp do` to `!$omp for` a while back, I think that we should raise an error when we encounter OpenMP pragmas that do not fit our grammar.

EmilyBourne commented:
@yguclu 
> In the PR description you say "Ignore any !$omp end do statements provided by the user", but I am not sure that I understand what you mean.
> 
> Since we changed our syntax from `!$omp do` to `!$omp for` a while back, I think that we should raise an error when we encounter OpenMP pragmas that do not fit our grammar.

Sorry this was a typo that I have now updated. It should be "Ignore any `!$omp end for` statements provided by the user". This is done for example [here](https://github.com/pyccel/pyccel/blob/master/tests/internal/scripts/openmp/ex2.py). I discussed the usefulness of this with @nhamidn and he felt that it could be easier for users who are used to using openmp in languages (ie fortran) where end statements are necessary. If an end statement is provided a warning is raised to indicate that it is ignored

nhamidn commented:
> @yguclu
> 
> > In the PR description you say "Ignore any !$omp end do statements provided by the user", but I am not sure that I understand what you mean.
> > Since we changed our syntax from `!$omp do` to `!$omp for` a while back, I think that we should raise an error when we encounter OpenMP pragmas that do not fit our grammar.
> 
> Sorry this was a typo that I have now updated. It should be "Ignore any `!$omp end for` statements provided by the user". This is done for example [here](https://github.com/pyccel/pyccel/blob/master/tests/internal/scripts/openmp/ex2.py). I discussed the usefulness of this with @nhamidn and he felt that it could be easier for users who are used to using openmp in languages (ie fortran) where end statements are necessary. If an end statement is provided a warning is raised to indicate that it is ignored

you still writing it in the wrong way, it's `#$omp `not `!$omp` in python

yguclu commented on pyccel/ast/omp.py:
```diff
@@ -76,6 +76,11 @@ def __getnewargs__(self):
         args = (self.txt, self.combined)
         return args
 
+    def __str__(self):
+        instruction = [self.name, self.combined, self.txt]
+        instruction = ' '.join(i for i in instruction if i)
+        return '#$ omp '+instruction
```
Sorry for being pedantic. Unless strictly necessary I would like us to avoid redefining variables, especially if their type changes as a result. Here the last two lines can be merged into one, and I think that the resulting code would be clearer.

EmilyBourne replied:
So you prefer:
```
instruction = ' '.join(i for i in [self.name, self.combined, self.txt] if i)
```
?
I disagree that this is clearer. I can rename the list variable if you prefer?

yguclu replied:
No, I mean:
```python
    def __str__(self):
        instruction = (self.name, self.combined, self.txt)
        return '#$ omp ' + ' '.join(i for i in instruction if i)
 ```

yguclu replied:
And yes, the name `instruction` is not really appropriate for a tuple...

EmilyBourne replied:
Ah of course sorry. Fixed: 20dfcbde

yguclu commented on pyccel/parser/grammar/openmp.tx:
```diff
@@ -42,11 +42,11 @@ OmpConstructOrDirective:
 ////////////////////////////////////////////////////
 OmpParallelConstruct:   'parallel'        combined=ParallelCombined? clauses*=AllOmpClauses;
 OmpLoopConstruct:       'for'             combinedsimd='simd'? clauses*=AllOmpClauses;
-OmpTaskLoopConstruct:   name='taskloop'   clauses*=AllOmpClauses;
+OmpTaskLoopConstruct:   'taskloop'   clauses*=AllOmpClauses;
 OmpSingleConstruct:     'single'          clauses*=AllOmpClauses;
 OmpCriticalConstruct:   name='critical'   clauses*=AllOmpClauses;
 OmpMaskedConstruct:     'masked'          combined=MaskedCombined? clauses*=AllOmpClauses;
-OmpSimdConstruct:       name='simd'       clauses*=AllOmpClauses;
+OmpSimdConstruct:       'simd'       clauses*=AllOmpClauses;
 OmpAtomicConstruct:     name='atomic'     clauses*=AllOmpClauses; // Not all clauses were added (can't test them).
 OmpTaskConstruct:       name='task'       clauses*=AllOmpClauses;
 OmpCancelConstruct:     name='cancel'     clauses*=OmpCancelType;
```
Could somebody explain why some constructs use `name='STRING'` and others just `'STRING'`?

EmilyBourne replied:
@nhamidn

nhamidn replied:
Constructs that don't accept any clauses like `#$omp master`, if it's grammar is : `OmpMasterConstruct:   'master'; ` will generate an error at this line https://github.com/pyccel/pyccel/blob/1621df5c8786fdaba5ec49b956a0ec980bb9c777/pyccel/parser/syntax/openmp.py#L863  saying `AttributeError: 'str' object has no attribute 'expr'`, and this is because textx will not create any object for this construct since we don't save any value in the grammar and will just return a string.

the `name= ` can be removed from the construct that save at least one value different than `name`.

yguclu commented on pyccel/parser/syntactic.py:
```diff
@@ -1034,6 +1034,7 @@ def _visit_CommentMultiLine(self, stmt):
 
                 txt = com[1:].lstrip()
                 exprs.append(Comment(txt))
+            exprs[-1].set_fst(stmt)
```
Is this a bug fix?

EmilyBourne replied:
A very small one yes. I have added it to the PR description

yguclu reviewed (changes requested):
Good job @EmilyBourne! I have a couple of comments and questions.
